### PR TITLE
rfc: introduce max partition policy to oneDNN Graph API

### PR DESCRIPTION
**Pull request by ElaineBao**
# Description

This RFC is to introduce max partition policy to oneDNN Graph API.

Rendered version: [link](https://github.com/ElaineBao/oneDNN/tree/yixin/rfc/max_partition_policy/rfcs/20230801-max-partition-policy)

### CODE REVIEW

File: rfcs/20230801-max-partition-policy/README.md [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+   on partitioning. This on the other hand means users may lose fine-grained control
+   of what optimizations they want to get from the library.
+1. The performance of max policy may or may not be better than fusion policy. The
+   library tries to provide the best performance by model performance heuristics.
+   But for untested models, it is possible that the performance of max policy is
+   worse than fusion policy.
+
+## Known issues
+
+1. DNNL backend will not support max partition policy in oneDNN v3.3 release.
```
Comment (niuxiaog) [inline rfcs/20230801-max-partition-policy/README.md:109]: When max partition policy is chosen by user, graph compiler will use it, even though other backends may not support this policy.

Comment (ElaineBao) [reply rfcs/20230801-max-partition-policy/README.md:109]: Yes, that's expected

**Review by niuxiaog — APPROVED** [review]

File: rfcs/20230801-max-partition-policy/README.md [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    ...
+};
+```
+
+The expected behaviors when using max policy:
+
+- If not specified, the default partition policy remains to be fusion policy.
+- Max policy returns partitions max in partition size. The criteria of partitioning
+  is based on pre-defined ops list and pre-defined rules. Connected ops
+  that meet the conditions of pre-defined ops list and pre-defined rules
```
Comment (chuanqi129) [inline rfcs/20230801-max-partition-policy/README.md:93]: I think this sentence may be ambiguous, `meet the condistions of pre-defined ops list and don't hit the pre-defined excluding rules will be selected into one partition` ?

File: rfcs/20230801-max-partition-policy/README.md
Comment (chuanqi129) [inline rfcs/20230801-max-partition-policy/README.md:1]: Will we have env var to change the pattern policy at the runtime?

**Review by chuanqi129 — APPROVED** [review]

File: rfcs/20230801-max-partition-policy/README.md (hunk 2 above)
Comment (ElaineBao) [reply rfcs/20230801-max-partition-policy/README.md:93]: `meet the conditions of pre-defined ops list and pre-defined rules will be selected into one partition` 
means both pre-defined ops list and pre-defined rules need to be met to be selected into one partition, violate either will not be selected into one partition.

File: rfcs/20230801-max-partition-policy/README.md
Comment (ElaineBao) [reply rfcs/20230801-max-partition-policy/README.md:1]: We don't support env var to change the partition policy, as it only used once when get_partition(), after partitions are formed, the policy will not change the partitions any more.

File: rfcs/20230801-max-partition-policy/README.md (hunk 2 above)
Comment (chuanqi129) [reply rfcs/20230801-max-partition-policy/README.md:93]: Yes, but i remember the rules are negative list. If op met one of rules, the op will be remove from the partition

File: rfcs/20230801-max-partition-policy/README.md [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        debug = dnnl_graph_partition_policy_debug,
+    };
+    ...
+};
+```
+
+The expected behaviors when using max policy:
+
+- If not specified, the default partition policy remains to be fusion policy.
+- Max policy returns partitions max in partition size. The criteria of partitioning
```
Comment (igorsafo) [inline rfcs/20230801-max-partition-policy/README.md:91]: What is motivation for differentiation between max and fusion? How user can decide which one should be used (not for a single model, but for overall integration)? Maybe "max" should be renamed to something like "experimental" ("experimental" is really bad word, we already have semantics for experimental features) to indicate that user shouldn't use it until they know it brings performance for their particular case?

Comment (ElaineBao) [reply rfcs/20230801-max-partition-policy/README.md:91]: > What is motivation for differentiation between max and fusion?

From my understanding, fusion defines fixed number of patterns, while max's optimizations are not fixed, it will generate different optimization for different models based on model architecture, which might be good for OOB models. 

> How user can decide which one should be used (not for a single model, but for overall integration)?

It depends on how much control user wants to give to the library. Compared with fusion, max gives the library more control to optimize the graph.
But I can understand user wants to get better performance when they give the library more control, which is not guaranteed for max policy since we are not able to test all the models in the world (Please correct me if I'm wrong @ZhennanQin @niuxiaog).

> Maybe "max" should be renamed to something like "experimental" ("experimental" is really bad word, we already have semantics for experimental features) to indicate that user shouldn't use it until they know it brings performance for their particular case? Another option to call it "all" if the intend is to take all graph and merge it into a single partition.

`all` may not be true because `take all graph and merge it into a single partition` is not guaranteed. `experimental` sounds a doable option, what do you think? @ZhennanQin @TaoLv

File: rfcs/20230801-max-partition-policy/README.md (hunk 2 above)
Comment (ElaineBao) [reply rfcs/20230801-max-partition-policy/README.md:93]: Oh I got your point. But since the pre-defined rules are backend specific, we cannot say it's always a negative list. So here I just keep general words.

File: rfcs/20230801-max-partition-policy/README.md
Comment (ElaineBao) [reply rfcs/20230801-max-partition-policy/README.md:1]: Yes, I think so

File: rfcs/20230801-max-partition-policy/README.md (hunk 3 above)
Comment (ZhennanQin) [reply rfcs/20230801-max-partition-policy/README.md:91]: `max` policy is designed to support direct programming with oneDNN graph API to implement custom ops or custom fusion patterns. For example, end-user can build RMS_norm with small ops and pass it to onednn graph with max policy, and compiler backend can generate a fused kernel for this. This is frequent request for newly introduced activation or normalization. I think `experimental` is not a good name since it can mean anything and doesn't reflect actual behavior. But we can use other term than `max` to make it more clear.

File: rfcs/20230801-max-partition-policy/README.md (hunk 1 above)
Comment (TaoLv) [reply rfcs/20230801-max-partition-policy/README.md:109]: >> even though other backends may not support this policy.

Need to clarify what "may not support" means here - should the backend just skip the backend directly or fallback to fusion policy or debug policy implicitly?

File: rfcs/20230801-max-partition-policy/README.md (hunk 2 above)
Comment (TaoLv) [reply rfcs/20230801-max-partition-policy/README.md:93]: Will the pre-defined ops list and pre-defined rules be expose through API or API documentation? If not, how does the user know when to use the max policy? Or are you suggesting users to always specify max policy for their models?

File: rfcs/20230801-max-partition-policy/README.md [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+1. The performance of max policy may or may not be better than fusion policy. The
+   library tries to provide the best performance by model performance heuristics.
+   But for untested models, it is possible that the performance of max policy is
+   worse than fusion policy.
+
+## Known issues
+
+1. DNNL backend will not support max partition policy in oneDNN v3.3 release.
+   So when running with a combination of max partition policy and DNNL backend,
+   max partition policy will fallback to fusion partition policy.
```
Comment (TaoLv) [inline rfcs/20230801-max-partition-policy/README.md:111]: Though I'm not a big fan for implicit fallback, it seems to be the only solution for now to retain performance for max policy vs. fusion policy. Please clarify if this is short-term or a long-term solution even after DNNL backend supports the policy. And also clarify if this decision applies to all backends of the library.

File: rfcs/20230801-max-partition-policy/README.md (hunk 1 above)
Comment (ElaineBao) [reply rfcs/20230801-max-partition-policy/README.md:109]: For now, there are two backends, dnnl backend and graph compiler backend. 
- Graph compiler backend will support max policy
- Dnnl backend will not. The behavior has been clarified as:
> when running with a combination of max partition policy and DNNL backend,
   max partition policy will fallback to fusion partition policy.

File: rfcs/20230801-max-partition-policy/README.md [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+1. The validation for max partition policy will not be landed in oneDNN v3.3
+   release, so the implementation of max partition policy in Graph Compiler
+   backend is a experimental feature.
```
Comment (TaoLv) [inline rfcs/20230801-max-partition-policy/README.md:114]: Though validation cannot be covered in v3.3, it's still needed to describe the validation methodology and potential options here. I see @chuanqi129 is suggesting to add more arguments to benchdnn command line. It would be better to include that into this RFC also.

File: rfcs/20230801-max-partition-policy/README.md (hunk 1 above)
Comment (TaoLv) [reply rfcs/20230801-max-partition-policy/README.md:109]: DNNL backend is not specific. The design decision will be applied to all the backends. See my comment at https://github.com/oneapi-src/oneDNN/pull/1694#discussion_r1285387269 .

File: rfcs/20230801-max-partition-policy/README.md (hunk 2 above)
Comment (ElaineBao) [reply rfcs/20230801-max-partition-policy/README.md:93]: > Will the pre-defined ops list and pre-defined rules be expose through API or API documentation?

Fusion policy doesn't expose fusion patterns through API, so I think max policy follows similar strategy. For document, maybe we can provide a page like [pre-defined fusion patterns](https://oneapi-src.github.io/oneDNN/dev_guide_graph_fusion_patterns.html), if it helps.

> If not, how does the user know when to use the max policy? Or are you suggesting users to always specify max policy for their models?

I feel this discussion can be merged with https://github.com/oneapi-src/oneDNN/pull/1694/files#r1283768894

File: rfcs/20230801-max-partition-policy/README.md (hunk 4 above)
Comment (ElaineBao) [reply rfcs/20230801-max-partition-policy/README.md:111]: This is a long-term solution: for all backends that don't support max policy, it will fallback to fusion policy. And if a backend doesn't support fusion policy, it will fallback to debug policy.

I will add above statement in the rfc.

File: rfcs/20230801-max-partition-policy/README.md (hunk 5 above)
Comment (ElaineBao) [reply rfcs/20230801-max-partition-policy/README.md:114]: good suggestion, will add.

**Comment by yimeisun123** [comment]
1) In the current oneDNN graph implementation, my understanding is that the graph compiler backend already provides additional fusions comparing to DNNL backend. Will specifying "max" policy bring in even more fusions in the graph compiler backend? What happens if we are using graph compiler backend without specifying max policy?
2) It would be good to know how you decide the op list and policy.
3) what would be the case that specifying max policy may have negative impact?

**Comment by niuxiaog** [comment]
> 1. In the current oneDNN graph implementation, my understanding is that the graph compiler backend already provides additional fusions comparing to DNNL backend. Will specifying "max" policy bring in even more fusions in the graph compiler backend? What happens if we are using graph compiler backend without specifying max policy?
> 2. It would be good to know how you decide the op list and policy.
> 3. what would be the case that specifying max policy may have negative impact?

1. Yes, specifying "max" policy will bring in even more fusions in the graph compiler backend. If we are using graph compiler backend without specifying `max` policy, the default `fusion` policy will be used.
2. The op list is roughly the supported ops list of graph compiler.
3. Models in the model zoo will get performance improvement by `max` policy; For untested models, there may be some regression.

**Review by yucai-intel — APPROVED** [review]

**Comment by TaoLv** [comment]
@niuxiaog 

> The op list is roughly the supported ops list of graph compiler.

That's internal. Users don't know which ops are supported by graph compiler from API.

> Models in the model zoo will get performance improvement by max policy; For untested models, there may be some regression.

I believe framework integration is for general usage. How can frameworks or framework users know which model is from model zoo and which is not? @chunyuan-w @sanchitintel @yucai-intel @retonym @wenzhe-nrv Please share how you plan to integrate and expose this feature. Thanks.

File: rfcs/20230801-max-partition-policy/README.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+## Proposal
+
+As mentioned in previous sessions, fusion policy relies on predefined fusion
+patterns to do partitioning, which may miss some optimization opportunities
+for customized ops and patterns that have not been predefined.
+To address this, max policy is proposed.
+
+The expected behaviors when using max policy:
+
+- Max policy relies on predefined ops list to do partitioning. Connected ops
```
Comment (vpirogov) [inline rfcs/20230801-max-partition-policy/README.md:79]: Choice between `max` and `fusion` policy falls on the shoulders of the framework user. How this is expected to be exposed in Pytorch and Tensorflow?

How users are supposed to decide whether use fusion or max? The explanation here talks about implementation details mainly, but what are the consequences visible to framework user?

File: rfcs/20230801-max-partition-policy/README.md [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+## Known issues
+
+1. For all backends that don't support max partition policy, they will fallback to
+   fusion partition policy implicitly.
+   Likewise, for all backends that don't support fusion partition policy, they
+   will fallback to debug partition policy implicitly.
+   For example, DNNL backend will not support max partition policy in oneDNN v3.3
+   release, so when running with a combination of max partition policy and DNNL
+   backend, max partition policy will fallback to fusion partition policy.
+1. The validation for max partition policy will not be landed in oneDNN v3.3
```
Comment (vpirogov) [inline rfcs/20230801-max-partition-policy/README.md:144]: We are not landing features without validation in oneDNN.

File: rfcs/20230801-max-partition-policy/README.md (hunk 6 above)
Comment (ZhennanQin) [reply rfcs/20230801-max-partition-policy/README.md:79]: My understanding is,

`Max` policy is not expected to be exposed by framework, instead, it's used to support framework developer to implement their custom fusion kernels.

`Fusion` policy is the recommended way to use onednn graph automatically, as we will do graph partition according to pre-defined fusion pattern and deliver good performance. While `max` policy is to support direct programming with onednn graph API, and framework developers need to program the graph they want to fuse and pass it to onednn graph with `max` policy, just like how they do C program and compile it with GCC. 

`max` policy can be used in below scenarios:

1. The default `fusion` policy doesn't meet framework's requirement and developer wants to change the fusion behavior, for example, for a 3-layer MLP, `fusion` policy will take them all, but if the developer only wants oneDNN graph to take the first two layers, and leave the last layer to be un-partitioned, then they can program with `max` policy and define a graph only contains first two layers.
2. Define new fusion patterns. For example, if framework developer want to implement RMSNorm op, they can build a graph with small ops and pass it to oneDNN graph with `max` policy, then a fused kernel will be returned.

In general, `max` policy is not designed for framework users, but for framework developers.

**Comment by sanchitintel** [comment]
Hi @ElaineBao, can framework developers tell if custom patterns that don't already exist with `fusion` policy (but can be added by a backend as predefined patterns) would perform better or worse than `max` policy? In other words, would `max` policy enable better OOB support by not having to request backend teams to add specific pre-defined patterns for new OOB models that would be encountered in the future? Thanks!

**Comment by ElaineBao** [comment]
Hi, @sanchitintel, I guess you are asking how a FW could determine when to use max and when to use fusion policy.

The general rule is that the default recommended policy is still fusion policy. But for a specific workload that have customed patterns that have not been covered in fusion policy, FW can try to use max policy to see if it achieves better performance.

File: rfcs/20230801-max-partition-policy/README.md (hunk 6 above)
Comment (sanchitintel) [reply rfcs/20230801-max-partition-policy/README.md:79]: @Jianhui-Li, adding to your point (custom ops vs. patterns created by max partitioning), I feel like there's another aspect as well - a pattern added manually by backend teams may sometimes perform better than the pattern automatically chosen by the max partitioning approach, and sometimes, the opposite may be true.

 So at runtime, should a framework (transparently to framework users) benchmark both approaches, and eventually use the approach that delivers better performance? In case a framework can't do this sort of benchmarking, then the framework may have to expose an API to the users to manually check if max partitioning approach (if it's non-default) may deliver better performance.

**Comment by TaoLv** [comment]
The proposal was reviewed internally and will not be implemented in the library as the use scenario in frameworks is not clear. Thank you all for the discussions and participation.
